src/constant_value.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use rustc::hir::def_id::DefId;
+use std::collections::HashMap;
+use summaries::PersistentSummaryCache;
+
+/// Abstracts over constant values referenced in MIR and adds information
+/// that is useful for the abstract interpreter. More importantly, this
+/// value can be serialized to the persistent summary cache.
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum ConstantValue {
+    /// A reference to a function
+    Function {
+        /// Indicates if the function is known to be treated specially by the Rust compiler
+        is_intrinsic: bool,
+        /// The key to use when retrieving a summary for the function from the summary cache
+        summary_cache_key: String,
+        // todo: is there some way store the def_id here if available?
+        // this would not be serialized/deserialized.
+    },
+    // A place holder for other kinds of constants. Eventually this goes away.
+    Unimplemented,
+}
+
+impl ConstantValue {
+    /// Returns a constant value that is a reference to a function
+    pub fn for_function(
+        def_id: DefId,
+        summary_cache: &mut PersistentSummaryCache,
+    ) -> ConstantValue {
+        let summary_cache_key = summary_cache.get_summary_key_for(def_id);
+        // todo: include the def_id in the result
+        ConstantValue::Function {
+            is_intrinsic: false,
+            summary_cache_key: summary_cache_key.to_owned(),
+        }
+    }
+}
+
+/// Keeps track of MIR constants that have already been mapped onto ConstantValue instances.
+pub struct ConstantValueCache {
+    cache: HashMap<DefId, ConstantValue>,
+    std_intrinsics_unreachable_function: Option<ConstantValue>,
```

Comment:
Will we need a new field for each intrinsic function we care about?

Reply:
I don't think so. Most of them can be modeled by summaries. It is possible that unreachable itself can be modeled by a  summary with a false precondition, but it will take quite a bit more infrastructure to be present before that will be actually useful. So, this is probably temporary scaffolding.